crates/net/src/http/response.rs

```diff
@@ -0,0 +1,268 @@
+    }
+
+    /// Gets the headers.
+    pub fn headers(&self) -> Headers {
+        Headers::from_raw(self.0.headers())
+    }
+
+    /// Has the response body been consumed?
+    ///
+    /// If true, then any future attempts to consume the body will error.
+    pub fn body_used(&self) -> bool {
+        self.0.body_used()
+    }
+
+    /// Gets the body.
+    pub fn body(&self) -> Option<web_sys::ReadableStream> {
+        self.0.body()
+    }
+
+    /// Reads the response to completion, returning it as `FormData`.
+    pub async fn form_data(&self) -> Result<web_sys::FormData, Error> {
+        let promise = self.0.form_data().map_err(js_to_error)?;
+        let val = JsFuture::from(promise).await.map_err(js_to_error)?;
+        Ok(web_sys::FormData::from(val))
+    }
+
+    /// Reads the response to completion, parsing it as JSON.
+    #[cfg(feature = "json")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
+    pub async fn json<T: DeserializeOwned>(&self) -> Result<T, Error> {
+        serde_json::from_str::<T>(&self.text().await?).map_err(Error::from)
+    }
+
+    /// Reads the response as a String.
+    pub async fn text(&self) -> Result<String, Error> {
+        let promise = self.0.text().unwrap();
+        let val = JsFuture::from(promise).await.map_err(js_to_error)?;
+        let string = js_sys::JsString::from(val);
+        Ok(String::from(&string))
+    }
+
+    /// Gets the binary response
+    ///
+    /// This works by obtaining the response as an `ArrayBuffer`, creating a `Uint8Array` from it
+    /// and then converting it to `Vec<u8>`
+    pub async fn binary(&self) -> Result<Vec<u8>, Error> {
+        let promise = self.0.array_buffer().map_err(js_to_error)?;
+        let array_buffer: ArrayBuffer = JsFuture::from(promise)
+            .await
+            .map_err(js_to_error)?
+            .unchecked_into();
+        let typed_buff: Uint8Array = Uint8Array::new(&array_buffer);
+        let mut body = vec![0; typed_buff.length() as usize];
+        typed_buff.copy_to(&mut body);
+        Ok(body)
+    }
+}
+
+impl From<web_sys::Response> for Response {
+    fn from(raw: web_sys::Response) -> Self {
+        Self(raw)
+    }
+}
+
+impl From<Response> for web_sys::Response {
+    fn from(res: Response) -> Self {
+        res.0
+    }
+}
+
+impl fmt::Debug for Response {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Response")
+            .field("url", &self.url())
+            .field("redirected", &self.redirected())
+            .field("status", &self.status())
+            .field("headers", &self.headers())
+            .field("body_used", &self.body_used())
+            .finish()
+    }
+}
+
+/// A writable wrapper around `web_sys::Reponse`: an http response to be used with the `fetch` API
+/// on a server side javascript runtime
+pub struct ResponseBuilder {
+    headers: Headers,
+    options: web_sys::ResponseInit,
+    body: Option<ResponseBody>,
```

Comment:
I don't think the body field needs to be saved. We can get a smaller struct size by removing the body method.